### PR TITLE
[stdlib] Add `repr()` function and `Representable` trait

### PR DESCRIPTION
**Pull request by gabrieldemarmiesse**
This PR is a follow up on the discussion there:

https://github.com/modularml/mojo/issues/2207

There is no `RepresentableRaising` trait yet, this can come later on.

We assume here that `42` is the Mojo representation of `Int(42)` because this is how it materializes. Same we assume that `"hello"` is the Mojo string representation of `String("hello")` because StringLiteral materializes into String.

We could have used `Int(42)` and `String("hello")` as Mojo representations but that seems very cumbersome and not aligned with the evolution of the language. The discussion is still open for other types.

The `__repr__` implementation for String is very simplistic and I believe it's enough as a start, we can push the complexity in other separate PRs (handling `\` and so on). I will open an issue when this PR is merged requesting contributor help to improve it.

I will create another issue after this PR is merged to track the conformance of (all?) library types to the `Representable` trait.

### CODE REVIEW

**Comment by jayzhan211** [comment]
Is `Representable` equivalent to `Stringable`?

File: stdlib/src/builtin/repr.mojo [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+        Returns:
+            The string representation of the instance.
+        """
+        pass
+
+
+fn repr[T: Representable](value: T) -> String:
+    """
+    Args:
```
Comment (JoeLoser) [inline stdlib/src/builtin/repr.mojo:67]: The doc validation fails on this - I recommend running the commands from https://github.com/modularml/mojo/pull/2344 locally until we have API doc string validation on the public CI.

Comment (gabrieldemarmiesse) [reply stdlib/src/builtin/repr.mojo:68]: Fixed it :)

**Comment by JoeLoser** [comment]
✅🟣 This contribution has been merged 🟣✅

Hey @gabrieldemarmiesse,

Thanks so much again for the contribution! 🎉

We're moving to a new infrastructure for merging contributions to Mojo (we're using a tool called [Copybara](https://github.com/google/copybara)), and your contribution has now been merged into our internal copy of the Mojo Standard Library. I've added the "merged-internally" label on this PR.

The changes in this PR will appear here in the mojo repo nightly branch when we do our next outbound synchronization at the time that the next Mojo nightly is released. That should happen within the next 24 hours.

Please let me know if you have any questions or concerns.

**Comment by gabrieldemarmiesse** [comment]
Thanks for merging and fixing the conflicts. I made two follow-up issues:
* https://github.com/modularml/mojo/issues/2458
* https://github.com/modularml/mojo/issues/2459

**Comment by JoeLoser** [comment]
🎉 🔥 This landed in tonight's nightly release in https://github.com/modularml/mojo/pull/2480!

Thanks again for the contribution!

**Comment by helehex** [comment]
another thing about `repr`, it takes non-printable characters and gives the representation of them, such as `\n`.
I already did this in one of my side projects, so I can add it to stdlib.

**Pull request closed without merging**